proposals/readonly-struct-init.md

```diff
@@ -0,0 +1,161 @@
+# readonly object initializers 
+
+* [x] Proposed
+* [ ] Prototype: Not Started
+* [ ] Implementation: Not Started
+* [ ] Specification: Below
+
+## Summary
+
+This feature will allow for assigning of `readonly` fields and get only auto-implemented properties
+in object initialezrs when the type being constructed is a `struct` or `ref struct`.
+
+## Motivation
+
+The intent of `readonly` in C# is to prevent the modification of a value after construction has 
+completed. That is once the code calling `new` has a reference to the value there should be no 
+further modification of `readonly` fields. Until the calling code has a reference to the object
+modifications can continue to be made via `this`.
+
+This makes it even easier to define usable `readonly struct` types in C#. The simplest form of
+declaring the fields now creates a fully usable value:
+
+``` csharp
+readonly struct Point { 
+    readonly int X;
+    readonly int Y; 
+}
+
+var p = new Point { X = 42; Y = 13 };
+```
+
+This means authoring `readonly struct` types in C# is just as easy as creating the less
+perferred mutable `struct` types. There is no longer a need to define a constructor which simply 
+re-lists every field as a parameter and then assigns them one at a time. The definition of the 
+field itself allows provides a mechanism for initialization
+
+## Detailed Design 
+
+### Language
+
+The language will extend the object initialization rules to allow for an accessible readonly field 
+or auto-implemented property to appear in the member initialization list. In such a context the 
+members will allow for assignment just as it would inside a constructor of the object. 
+
+The following restrictions will apply to this extension:
+
+1. The type being instantiated must be a `struct` or `ref struct`. 
+2. The constructor used in the initializer cannot have any `out` or `ref` parameters.
+
+The (2) restriction exists because such a parameter could smuggle out `this` from the object 
+constructor. That would make a reference to the constructed object visible to the code calling
+`new` and hence allow it to observe mutations to the value.
+
+### Compiler 
+
+The compiler will change how the backing field of an auto-implemented property is emitted in the 
+following ways:
+
+- The accessibility of the backing field will match the accessibility of the property.
+- The name of the field will be unspeakable (illegal C# identifier hence never imported) but it 
+will be possible to map such fields to the properties that defined them.
+
+Taken together this allows the C# compiler to translate a get only auto-implemented property into
+a simple field initialization when used as an object initializer. 
+
+``` csharp 
+struct Name {
+    // Backing field: public readonly string <FirstName>__BackingField;
+    public string FirstName { get; }
+    // Backing field: public readonly string <LastName>__BackingField;
```

Comment:
What will these be marked as in metadata? Special named?

This would now be a breaking change to move from auto-property -> normal property.

Reply:
I think the breaking change aspect negates the whole point of properties. Might as well limit this feature to fields IMHO.

Reply:
Also be a breaking change to compiler with an earlier compiler. Do we have many examples of that?

Reply:
> What will these be marked as in metadata? Special named?

Today they are marked as `[CompilerGenerated]` and `DebuggerBrowsable(false)`. The `specialname` IL modifier isn't applied here. I don't believe C# does that for any field, but it does for methods. I don't know the history here and / or the implications of that. 

> This would mean that it be a breaking change to move from auto-property -> normal property. I think this should be called out.

I will add that. There are already subtle breaking change issues in this area. For example converting a prop to an auto-implemented property can break struct member-wise initialization. But this is a new one AFAICT and will call it out. 

> Might as well limit this feature to fields IMHO.

I considered that but feel like benefits outweigh the downsides here. Want to continue exploring it for now.